### DISCUSSION TITLE
Window management: Handling clients that want attention

### DISCUSSION THREAD

**Discussion by vain** [labwc/labwc#1065] [Ideas]
I have a few questions about handling clients that want attention.

Consider this scenario:

-   Run Firefox on "Workspace 1".
-   Change to another workspace.
-   Run something like a mail client and open a URL from within that client.
-   labwc now switches back to "Workspace 1" and gives Firefox input focus.

I find this very irritating. I usually read mail or RSS feeds, open a bunch of links, and then go back to Firefox afterwards to read them in bulk.

In general, clients stealing focus is a bit annoying, *especially* if they're on another workspace.

The way Sway handles this is to keep the focus on the current client, but mark the Firefox window with something like a red border, which also marks the corresponding workspace in a pager. This leads me to the second part: "Urgency hints".

Basically, Firefox "wants attention". This is similar to "ringing the terminal bell", e.g. running `sleep 2; printf '\a'` in a terminal emulator like Foot (switch to another window during `sleep 2`). On X11, this used to set the window's "urgency hint", without doing anything related to input focus. I use the terminal bell a lot to alert me about unfocused programs when they have something to say or waiting on user input. (Currently, ringing the bell from within Foot does nothing on labwc but results in a highlighted window on Sway.)

How about implementing "the Sway way" in labwc? Don’t change input focus, but just mark the window and the workspace that it’s on in some way. Something along these lines:

```xml
<core>
    <onActivationRequest>
        <!-- Current behaviour: -->
        <action name="Focus"/>

        <!-- Alternatively, don't transfer focus, just mark client and workspace: -->
        <action name="MarkUrgent"/>
    </onActivationRequest>
</core>
```

Based on [this part of discussion #1002](https://github.com/labwc/labwc/discussions/1002#discussioncomment-6733502), I _suspect_ that all this comes from [XDG activation](https://wayland.app/protocols/xdg-activation-v1), but I'm very new to Wayland, so, no idea really. I also suspect that Sway is actually handling this protocol in a non-conforming way, since it doesn't actually activate the clients. So, if anything, my request would probably have to be implemented as a configuration setting anyway, I guess.

What are your thoughts on this? Is this on the roadmap, out of scope, working as intended, or did I misunderstand the whole thing? :)

(I realize that all this is just my personal workflow and maybe not a generic issue.)

(I also realize that the Openbox theme spec only includes `window.active` and `window.inactive`, not `window.urgent`, IIUC. The original Openbox made windows blink, i.e. switch between `window.active` and `window.inactive`. Openbox also transferred input focus like labwc, but only if the other client was on the same workspace.)

PS.: Thank you for your efforts on labwc! :)

**Comment by stefonarch** [comment]
Probably the  window-rule property `allowFocusRequest` would do the same?
https://github.com/labwc/labwc/issues/941

Personally I like the activation as in most cases I want activation, just in quiterss on X11 I used the option "open links in browser in background", similar to your workflow.

**Reply by vain** [reply]
That would probably solve the focus stealing issue with Firefox, yes. 🤔

(Doesn't help with "urgency hints". But maybe they're a separate issue after all. They're kind of related, so I lumped them together with the focus stuff, but yeah ...)

**Reply by Narrat** [reply]
Great, that this on the radar.
Biggest irk since I switched. Ever since speed switching workspaces because I open links from my RSS reader and didn't want to see the browser yet :D

**Comment by johanmalm** [comment]
Great question.

1. Allowing/disallowing activate-request is something we should do (whether it’s xdg-activation for Wayland clients or an X11 activate-request). I suggest we do this as a window-rule. I hadn’t previously thought about what you describe, but it feels like there are three modes here: activate, background and ignore.
2. I don't think we’ve discussed window.urgent theme options before. This is probably because we don’t support X11-urgency-hints (yet) and IIRC Wayland protocols don’t (yet) support urgency hints. So with a “background” mode as described in (1) it would make sense to add window.urgent. It begs the question if we should investigate X11 urgency hints also. I personally miss them for IRC.
3. Supporting urgency-hints on panels is a feature for the future because AFAIR there is not yet a protocol for neither urgency hints nor workspaces.

**Comment by johanmalm** [comment]
Just had a look - swaybar uses i3ipc to get the workspaces.

https://github.com/swaywm/sway/blob/89f85312687b7fd1777c936e169e6400cee0a4b4/swaybar/bar.c#L274

**Comment by Consolatis** [comment]
> In general, clients stealing focus is a bit annoying, especially if they're on another workspace.

To request valid tokens (that can be used to activate a window / steal the focus) the requesting application
- requires a user action like mouse click
- must currently own mouse or keyboard focus

This prevents a random application in the background to suddenly jump to the top.
Additionally integrating the target application which then uses such a token with a `allowFocusRequest` property sounds good.

Some background on our implementation:
- #758

And the protocol itself: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/staging/xdg-activation/xdg-activation-v1.xml

I've skimmed trough the protocol again and it doesn't seem to state anything about its double use function to serve as an "urgency" hint with invalid tokens but I must have read that somewhere. So technically instead of rejecting activation requests from invalid tokens we could indeed add some urgency state to the window that requested the activation (by default I think foot requests an invalid token and tries to activate itself using that token in the bell handler).

> Supporting urgency-hints on panels is a feature for the future because AFAIR there is not yet a protocol for neither urgency hints nor workspaces.

Maybe we should bring up the urgency-hint in https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/196

**Comment by vain** [comment]
Nice, I'm glad to hear that you consider working on this. :)

**Comment by Consolatis** [comment]
- #1067

Testing would be welcome.

**Comment by vain** [comment]
FTR, there is also this:

https://wayland.app/protocols/xdg-system-bell-v1

foot has implemented support for it in January:

https://codeberg.org/dnkl/foot/commit/2a07a2e6b9a9565287996c51ceb2824b37d3fba7

**Reply by Consolatis** [reply]
The protocol seems to be implemented in wlroots 0.19.x so we could technically use it.
